### ISSUE TITLE
Enable Container to connect as network gateway

### ISSUE THREAD

**Issue by bhyslop-serrano** [containers/podman#24920]
### Feature request description

I want a container running less trusted code ("BOTTLE") to be able to use a trusted, carefully configured container ("SENTRY") to intermediate all local and internet access.

Find at [RBM System Vision](https://scaleinv.github.io/recipebottle) a diagram (also attached below) and abstract of my passion open-source project that needs this.

Prototypes using podman with bespoke BOTTLE dockerfiles have proven the concept. However, for this project to succeed, it needs to be able to use existing images and/or stock dockerfiles for BOTTLE.

Could podman add a network feature to allow SENTRY to function as gateway to BOTTLE from its earliest DHCP?

![Image](https://github.com/user-attachments/assets/682812aa-552b-491f-a7e9-d7cd6ee80e3b)

### Suggest potential solution

Based on my experiments, I propose adding an `--as-gateway` flag to the `podman network connect` command.
This parameterless option would:

1. Assign the gateway IP to the specified container (SENTRY in my case)
2. Handle MAC address consistency to avoid ARP cache issues
3. Ensure proper network initialization during container startup

This simple, unobtrusive feature could enable many security-focused container patterns. A search of your open issues led me only to [DMZ Feature Request](https://github.com/containers/podman/issues/20222) which may be consonant, though idle.

The proposed flag might align well with podman's transition to netavark, potentially making implementation straightforward.

### Have you considered any alternatives?

My road to this feature request has been long. Since I'm not a deep networking expert, I've been freely using Claude from Anthropic and ChatGPT to help me work on this journey from docker to podman.

After finding that docker couldn't connect host and internal networks to the same container, I switched to podman and explored several approaches:

1. **Direct Gateway Assignment:** I first tried the naive approach - simply assigning the gateway IP to the SENTRY container. Podman silently rejected this request.
2. **Network Configuration Deep Dive:** Next came experimentation with --opt and --dns options in podman commands, along with AI-suggested CNI configuration nudges. These didn't just work.
3. **BOTTLE-side Solutions:** Implementing dhclient in BOTTLE showed promise but revealed network race conditions during container startup. Some test BOTTLEs worked, others didn't - and more importantly, this required intimate BOTTLE container startup modifications.
4. **Privileged BOTTLE Container Approach:** Success came with elevating BOTTLE privileges to allow network stack modification, but this violated the core security premise of using SENTRY to protect untrusted BOTTLE containers.  Race condition vulnerabilities here too.
5. **Post-startup SENTRY Reconfiguration:** My latest attempts focused on reassigning SENTRY's IP after startup. This led to fascinating podman machine network namespace investigation with tcpdump, but ultimately failed due to ARP cache complications (AIs themselves uncertain that gratuitous ARP cache poisoning would work).
6. **Future-proofing Investigation:** When ChatGPT suggested deeper focus on CNI customization, I learned podman is transitioning from CNI to netavark. I explored other container runtimes for stable future-proof CNI environments, from locally managed containerd to CRI-O. However, I think people who might want to use my open source starting point will want the full 'desktop' feature set absent from CRI-O and other kubernetes-focused alternatives.

### Additional context

Thank you podman maintainers for an amazing project!
I'm not averse to attempting an implementation PR myself, once we scrub this concept for consonance with your long term visions.

**Comment by bhyslop** [comment]
Whoops, I made the above request from a customer linked account.  This is my primary identity.  Thanks podman!

**Comment by Luap99** [comment]
I don't think it makes sense for us to support something like this. There is so much network functionality already, I cannot deal with much more.

Netavark also allows plugins https://github.com/containers/netavark/blob/main/plugin-API.md so you can write your own network driver that does most of what you want.

In general for namespace setup you can configure any netns however you like and attach it to a container with `--network ns:/path/to/netns`. Or use oci hooks to configure it.

Another options is to configure the sentry container however you like and the attach the other container directly to the namespace via `--network container:sentry`. That however means they share localhost and all ports. But it would not require any ip/routing setup for the second container. You could then have one sentry container per bottle container.


Moving the gateway somewhere else is not a simple ask and would require quite a lot of changes so I am going to close this as not planned.

**Comment by bhyslop** [comment]
Hi @Luap99 ,

I wanted to drop a line to thank you for your considerate response. A few days after it (once I'd processed disappointment that you aren't a genie who grants wishes!), I sat down with the then new ChatGPT `-o1` and thoroughly explored this feature request alongside your response options. This led me to realize I had misunderstood network namespaces - opening up a new path! After a considerable wallow phase, the project came roaring back to life.

While I may not publicly share all my AI learnings from this process, I'll share one with you: sometimes they can do amazing things once you lay out a solid path for them, and your suggestions proved to be exactly that. They needed the help, but once provided, this embedded developer was able to move the ball toward a long-held goal: a safer way to use containers.

I'll update you when I release the alpha version, should you be interested. In brief, the result is an asymmetric podman machine namespace configuration with script-driven construction of virtual network connections, attachment of one to SENTRY inherent process namespace via PID, bridge creation, and assignment of BOTTLE to the namespace of the other one. I'm almost sure it's working as intended :).

Currently it's implemented in makefiles and bash scripts with a GitHub action. The AIs are suggesting a golang rewrite, which I'm considering - development in bash/make hasn't been ideal. Are you aware of a forum where I could discuss this idea with security-focused individuals? The bash strategy with explicit iptables was chosen to allow network-aware people to assess by inspection. I'm unsure whether there's a community that can as readily inspect the implied podman API/golang/Netlink/nft/yaml implementation. This will wait until after the current version's release - it's time to get on the field.

Current work focuses on building out sufficient testing, addressing numerous 'fixme' aspects, and completing documentation. Unless you object, I'll acknowledge your help in the release - it was truly critical.

Your handling of this feature request demonstrates that the podman project is in the hands of serious, mission-focused people. This is tremendously reassuring, given podman's importance to this container project. Thank you!

Best regards,
Brad
